components/script/dom/offscreencanvas.rs

```diff
@@ -0,0 +1,178 @@
+use dom::bindings::reflector::reflect_dom_object;
+use dom::bindings::root::{DomRoot, Dom};
+use dom::bindings::str::DOMString;
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::htmlcanvaselement::HTMLCanvasElement;
+use dom::offscreencanvasrenderingcontext2d::OffscreenCanvasRenderingContext2D;
+use dom_struct::dom_struct;
+use euclid::Size2D;
+use js::rust::HandleValue;
+use js::jsapi::JSContext;
+use ref_filter_map;
+use std::cell::Ref;
+use std::cell::Cell;
+
+#[must_root]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+pub enum OffscreenCanvasContext {
+    OffscreenContext2d(Dom<OffscreenCanvasRenderingContext2D>),
+    //WebGL(Dom<WebGLRenderingContext>),
+    //WebGL2(Dom<WebGL2RenderingContext>),
+}
+
+
+#[dom_struct]
+pub struct OffscreenCanvas{
+    eventtarget: EventTarget,
+    height: Cell<u64>,
+    width: Cell<u64>,
+    context: DomRefCell<Option<OffscreenCanvasContext>>,
+    placeholder: Option<Dom<HTMLCanvasElement>>,
+}
+
+impl OffscreenCanvas{
+    pub fn new_inherited(height: u64, width: u64, placeholder: Option<&HTMLCanvasElement>) -> OffscreenCanvas {
+        OffscreenCanvas {
+            eventtarget: EventTarget::new_inherited(),
+            height: Cell::new(height),
+            width: Cell::new(width),
+            context: DomRefCell::new(None),
+            placeholder: placeholder.map(Dom::from_ref),
+        }
+    }
+
+    pub fn new(global: &GlobalScope, height: u64, width: u64, placeholder: Option<&HTMLCanvasElement>) -> DomRoot<OffscreenCanvas> {
+        reflect_dom_object(Box::new(OffscreenCanvas::new_inherited(height,width,placeholder)), global, OffscreenCanvasWrap)
+    }
+
+    pub fn Constructor (global: &GlobalScope, height: u64, width: u64) -> Fallible<DomRoot<OffscreenCanvas>> {
+        //step 1
+        let offscreencanvas = OffscreenCanvas::new(global,height,width,None);
+        //step 2
+
+        if offscreencanvas.context.borrow().is_some() {
+            return Err(Error::InvalidState);
+        }
+
+        //offscreencanvas.placeholder = None;
+
+        //step 3
+        Ok(offscreencanvas)
+    }
+
+    pub fn get_size(&self) -> Size2D<u64> {
+        Size2D::new(self.Width(), self.Height())
+    }
+
+    pub fn context(&self) -> Option<Ref<OffscreenCanvasContext>> {
+        ref_filter_map::ref_filter_map(self.context.borrow(), |ctx| ctx.as_ref())
+    }
+
+    #[allow(unsafe_code)]
+    fn get_or_init_2d_context(&self,cx: *mut JSContext) -> Option<DomRoot<OffscreenCanvasRenderingContext2D>> {
+        if let Some(ctx) = self.context() {
+            return match *ctx {
+                OffscreenCanvasContext::OffscreenContext2d(ref ctx) => Some(DomRoot::from_ref(ctx)),
+                _ => None,
+            };
+        }
+        //let global = window_from_node(self);
+        let size = self.get_size();
+        let context = unsafe {OffscreenCanvasRenderingContext2D::new(&GlobalScope::from_context(cx), self, size)};
```

Comment:
This can use `let context = OffscreenCanvasRenderingContext2D::new(self.global(), self, size);` instead.

Reply:
@jdm This statement give error as offscreencanvas does not have it's own global defined.
Error:"error[E0599]: no method named `global` found for type `&dom::offscreencanvas::OffscreenCanvas` in the current scope
"

Reply:
Any other alternatives?